nixos/modules/services/networking/crab-hole.nix

```diff
@@ -0,0 +1,352 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+with lib;
+let
+  cfg = config.services.crab-hole;
+  configFormat = pkgs.formats.toml { };
+  filename = file: lib.lists.last (builtins.split "/" (builtins.toString file));
+in
+{
+  options = {
+    services.crab-hole = {
+      enable = mkEnableOption "Crab-hole Service";
+
+      user = mkOption {
+        type = types.str;
+        default = "crab-hole";
+        description = "User account under which crab-hole runs.";
+      };
+
+      group = mkOption {
+        type = types.str;
+        default = "crab-hole";
+        description = "Group account under which crab-hole runs.";
+      };
+
+      workDir = mkOption {
+        type = types.str;
+        default = "/var/lib/crab-hole";
+        description = "Crab-holes data directory.";
+      };
+
+      config = mkOption {
+        default = null;
+        description = "Crab-holes config";
+
```

Comment:
An example wold be great. You can reuse [upstream example](https://github.com/LuckyTurtleDev/crab-hole/blob/main/example-config.toml) here using https://pseitz.github.io/toml-to-json-online-converter/ and https://json-to-nix.pages.dev

```suggestion
        example = {
          downstream = [
            {
              listen = "localhost";
              port = 8080;
              protocol = "udp";
            }
            {
              listen = "[::]";
              port = 8053;
              protocol = "udp";
            }
            {
              certificate = "dns.example.com.crt";
              key = "dns.example.com.key";
              listen = "[::]";
              port = 8054;
              protocol = "tls";
              timeout_ms = 3000;
            }
            {
              certificate = "dns.example.com.crt";
              dns_hostname = "dns.example.com";
              key = "dns.example.com.key";
              listen = "[::]";
              port = 8055;
              protocol = "https";
              timeout_ms = 3000;
            }
            {
              certificate = "dns.example.com.crt";
              dns_hostname = "dns.example.com";
              key = "dns.example.com.key";
              listen = "127.0.0.1";
              port = 8055;
              protocol = "quic";
              timeout_ms = 3000;
            }
          ];
          api = {
            admin_key = "1234";
            listen = "127.0.0.1";
            port = 8080;
            show_doc = true;
          };
          blocklist = {
            allow_list = [
              "file:///allowed.txt"
            ];
            include_subdomains = true;
            lists = [
              "https://raw.githubusercontent.com/StevenBlack/hosts/master/alternates/fakenews-gambling-porn/hosts"
              "https://s3.amazonaws.com/lists.disconnect.me/simple_tracking.txt"
              "file:///blocked.txt"
            ];
          };
          upstream = {
            name_servers = [
              {
                protocol = "tls";
                socket_addr = "[2606:4700:4700::1111]:853";
                tls_dns_name = "1dot1dot1dot1.cloudflare-dns.com";
                trust_nx_responses = false;
              }
              {
                protocol = "tls";
                socket_addr = "[2606:4700:4700::1001]:853";
                tls_dns_name = "1dot1dot1dot1.cloudflare-dns.com";
                trust_nx_responses = false;
              }
              {
                protocol = "tls";
                socket_addr = "1.1.1.1:853";
                tls_dns_name = "1dot1dot1dot1.cloudflare-dns.com";
                trust_nx_responses = false;
              }
              {
                protocol = "tls";
                socket_addr = "1.0.0.1:853";
                tls_dns_name = "1dot1dot1dot1.cloudflare-dns.com";
                trust_nx_responses = false;
              }
            ];
            options = {
              validate = true;
            };
          };
        };
```